apps/evm/lib/evm/message_call.ex

```diff
@@ -125,43 +125,78 @@ defmodule EVM.MessageCall do
     %{message_call | current_exec_env: exec_env}
   end
 
-  def update_state({gas_remaining, n_sub_state, n_exec_env, output}, message_call) do
-    if output == :failed do
-      failed_call(message_call)
-    else
-      {out_offset, _out_size} = message_call.output_params
-
-      machine_state =
-        message_call.current_machine_state
-        |> MachineState.push(1)
-        |> MachineState.refund_gas(gas_remaining)
-
-      machine_state =
-        if output == :invalid_input do
-          %{machine_state | last_return_data: []}
-        else
-          updated_machine_state = Memory.write(machine_state, out_offset, output)
-          list_output = :binary.bin_to_list(output)
-
-          %{updated_machine_state | last_return_data: list_output}
-        end
-
-      exec_env =
-        Map.put(message_call.current_exec_env, :account_interface, n_exec_env.account_interface)
-
-      sub_state =
-        message_call.current_sub_state
-        |> SubState.merge(n_sub_state)
-        |> SubState.add_touched_account(message_call.recipient)
-
-      %{
-        machine_state: machine_state,
-        exec_env: exec_env,
-        sub_state: sub_state
-      }
+  def update_state(
+        exec_result = {_gas_remaining, _n_sub_state, _n_exec_env, output},
+        message_call
+      ) do
+    case output do
+      :failed ->
+        failed_call(message_call)
+
+      {:revert, _output} ->
+        finalize_reverted_message_call(exec_result, message_call)
+
+      _ ->
+        finalize_successful_message_call(exec_result, message_call)
     end
   end
 
+  defp finalize_reverted_message_call(
+         {gas_remaining, _n_sub_state, _ren_exec_env, {:revert, output}},
+         message_call
+       ) do
+    {out_offset, _out_size} = message_call.output_params
+
+    machine_state =
+      message_call.current_machine_state
+      |> MachineState.push(0)
+      |> MachineState.refund_gas(gas_remaining)
+
+    updated_machine_state = Memory.write(machine_state, out_offset, output)
+    list_output = :binary.bin_to_list(output)
+
+    updated_machine_state = %{updated_machine_state | last_return_data: list_output}
+
+    %{
+      machine_state: updated_machine_state
+    }
+  end
+
+  defp finalize_successful_message_call(
+         {gas_remaining, n_sub_state, n_exec_env, output},
+         message_call
+       ) do
+    {out_offset, _out_size} = message_call.output_params
+
+    machine_state =
+      message_call.current_machine_state
+      |> MachineState.push(1)
```

Comment:
I'm not quite certain about the meaning of pushing a `0` vs a `1` when the operation fails vs succeeds. Do you think there would be a way to make that more clear? Maybe we could extract this `1` and the `0` in the `finalize_revert_message_call` into module attributes and name them?

Reply:
:+1: . Btw it's defined in operation's description in the Yellow Paper